File: src/__tests__/ControlFrame.test.ts
// WebSocketRT: Real-time websocket library
// Copyright (c) Leo C. Singleton IV <leo@leosingleton.com>
// See LICENSE in the project root for license information.

import { ControlFrame, DataFrameControl, MessageCancelControl } from '../ControlFrame';
import { TransportCapabilities, TransportCapabilities1 } from '../TransportCapabilities';

describe('ControlFrame', () => {

  it('Ensures a control frame respresenting capabilities (OpCode=0x00) can be serialized and deserialized', () => {
    const frame1 = new ControlFrame();
    frame1.opCode = 0x00;
    frame1.rttEstimate = 42;
    frame1.throughputEstimate = 12345678;
    frame1.frameData = new TransportCapabilities();
    frame1.frameData.majorVersion = 3;
    frame1.frameData.minorVersion = 5;
    frame1.frameData.capabilities1 = TransportCapabilities1.Capabilities | TransportCapabilities1.Capabilities2;
    const bytes = frame1.writeFrame();

    const frame2 = new ControlFrame();
    frame2.readFrame(bytes);
    expect(frame1.opCode).toEqual(frame2.opCode);
    expect(frame1.rttEstimate).toEqual(frame2.rttEstimate);
    expect(frame1.throughputEstimate).toEqual(frame2.throughputEstimate);
    const data1 = frame1.frameData;
    const data2 = frame2.frameData as TransportCapabilities;
    expect(data1.majorVersion).toEqual(data2.majorVersion);
    expect(data1.minorVersion).toEqual(data2.minorVersion);
    expect(data1.capabilities1).toEqual(data2.capabilities1);
  });

  it('Ensures a control frame respresenting a ping can be serialized and deserialized', () => {
    const frame1 = new ControlFrame();
    frame1.opCode = 0x10;
    frame1.rttEstimate = 42;
    frame1.throughputEstimate = 12345678;
    const bytes = frame1.writeFrame();

    const frame2 = new ControlFrame();
    frame2.readFrame(bytes);
    expect(frame1.opCode).toEqual(frame2.opCode);
    expect(frame1.rttEstimate).toEqual(frame2.rttEstimate);
    expect(frame1.throughputEstimate).toEqual(frame2.throughputEstimate);
    expect(frame1.frameData).toEqual(frame2.frameData);
  });

  it('Ensures a control frame preceding data frames can be serialized', () => {
    const frame1 = new ControlFrame();
    frame1.opCode = 0x02;
    frame1.rttEstimate = 4096;
    frame1.throughputEstimate = 123456789;
    frame1.frameData = [new DataFrameControl(), new DataFrameControl()];
    frame1.frameData[0].messageNumber = 4;
    frame1.frameData[0].dataOffset = 0;
    frame1.frameData[0].messageLength = 15000;
    frame1.frameData[0].isFirst = true;
    frame1.frameData[0].isLast = false;
    frame1.frameData[0].header = new Uint8Array([1, 2, 3, 4, 5, 6, 7]);
    frame1.frameData[1].messageNumber = 15;
    frame1.frameData[1].dataOffset = 19000000;
    frame1.frameData[1].messageLength = 20000000;
    frame1.frameData[1].isFirst = false;
    frame1.frameData[1].isLast = true;
    const bytes = frame1.writeFrame();

    const frame2 = new ControlFrame();
    frame2.readFrame(bytes);
    expect(frame1.opCode).toEqual(frame2.opCode);
    expect(frame1.rttEstimate).toEqual(frame2.rttEstimate);
    expect(frame1.throughputEstimate).toEqual(frame2.throughputEstimate);
    expect(frame1.frameData).toEqual(frame2.frameData as DataFrameControl[]);
  });

  it('Ensures a control frame respresenting a cancel (OpCode=0x12) can be serialized and deserialized', () => {
    const frame1 = new ControlFrame();
    frame1.opCode = 0x12;
    frame1.rttEstimate = 4096;
    frame1.throughputEstimate = 123456789;
    frame1.frameData = new MessageCancelControl();
    frame1.frameData.messageNumbers = 42;
    const bytes = frame1.writeFrame();

    const frame2 = new ControlFrame();
    frame2.readFrame(bytes);
    expect(frame1.opCode).toEqual(frame2.opCode);
    expect(frame1.rttEstimate).toEqual(frame2.rttEstimate);
    expect(frame1.throughputEstimate).toEqual(frame2.throughputEstimate);
    expect(frame1.frameData).toEqual(frame2.frameData as MessageCancelControl);
  });

});

File: dotnet/UnitTests/ControlFrameTest.cs
// WebSocketRT: Real-time websocket library
// Copyright (c) Leo C. Singleton IV <leo@leosingleton.com>
// See LICENSE in the project root for license information.

using Xunit;

namespace LeoSingleton.WebSocketRT.UnitTests
{
    public class ControlFrameTest
    {
        /// <summary>
        /// Ensures a control frame respresenting capabilities (OpCode=0x00) can be serialized and deserialized
        /// </summary>
        [Fact]
        public void SerializeDeserializeCapabilities()
        {
            var frame1 = new ControlFrame()
            {
                OpCode = 0x00,
                RttEstimate = 42,
                ThroughputEstimate = 12345678,
                Capabilities = new TransportCapabilities()
                {
                    MajorVersion = 3,
                    MinorVersion = 5,
                    Capabilities1 = TransportCapabilities1.Capabilities | TransportCapabilities1.Capabilities2
                }
            };
            var bytes = frame1.Write();

            var frame2 = new ControlFrame(bytes);
            Assert.Equal(frame1.OpCode, frame2.OpCode);
            Assert.Equal(frame1.RttEstimate, frame2.RttEstimate);
            Assert.Equal(frame1.ThroughputEstimate, frame2.ThroughputEstimate);
            Assert.Equal(frame1.Capabilities.MajorVersion, frame2.Capabilities.MajorVersion);
            Assert.Equal(frame1.Capabilities.MinorVersion, frame2.Capabilities.MinorVersion);
            Assert.Equal(frame1.Capabilities.Capabilities1, frame2.Capabilities.Capabilities1);
            Assert.Equal(frame1.DataFrames, frame2.DataFrames);
            Assert.Equal(frame1.CancellationDetails, frame2.CancellationDetails);
        }

        /// <summary>
        /// Ensures a control frame respresenting a ping can be serialized and deserialized
        /// </summary>
        [Fact]
        public void SerializeDeserializePing()
        {
            var frame1 = new ControlFrame()
            {
                OpCode = 0x10,
                RttEstimate = 42,
                ThroughputEstimate = 12345678
            };
            var bytes = frame1.Write();

            var frame2 = new ControlFrame(bytes);
            Assert.Equal(frame1.OpCode, frame2.OpCode);
            Assert.Equal(frame1.RttEstimate, frame2.RttEstimate);
            Assert.Equal(frame1.ThroughputEstimate, frame2.ThroughputEstimate);
            Assert.Equal(frame1.Capabilities, frame2.Capabilities);
            Assert.Equal(frame1.DataFrames, frame2.DataFrames);
            Assert.Equal(frame1.CancellationDetails, frame2.CancellationDetails);
        }

        /// <summary>
        /// Ensures a control frame preceding data frames can be serialized
        /// </summary>
        [Fact]
        public void SerializeDeserializeDataFrames()
        {
            var frame1 = new ControlFrame()
            {
                OpCode = 0x02,
                RttEstimate = 4096,
                ThroughputEstimate = 123456789,
                DataFrames = new DataFrameControl[]
                {
                    new DataFrameControl()
                    {
                        MessageNumber = 4,
                        Offset = 0,
                        Length = 15000,
                        IsFirst = true,
                        IsLast = false,
                        Header = new byte[] { 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07 }
                    },
                    new DataFrameControl()
                    {
                        MessageNumber = 15,
                        Offset = 19000000,
                        Length = 20000000,
                        IsFirst = false,
                        IsLast = true
                    }
                }
            };
            var bytes = frame1.Write();

            var frame2 = new ControlFrame(bytes);
            Assert.Equal(frame1.OpCode, frame2.OpCode);
            Assert.Equal(frame1.RttEstimate, frame2.RttEstimate);
            Assert.Equal(frame1.ThroughputEstimate, frame2.ThroughputEstimate);
            Assert.Equal(frame1.Capabilities, frame2.Capabilities);
            Assert.Equal(frame1.DataFrames.Length, frame2.DataFrames.Length);
            Assert.Equal(frame1.DataFrames[0].MessageNumber, frame2.DataFrames[0].MessageNumber);
            Assert.Equal(frame1.DataFrames[0].Offset, frame2.DataFrames[0].Offset);
            Assert.Equal(frame1.DataFrames[0].Length, frame2.DataFrames[0].Length);
            Assert.Equal(frame1.DataFrames[0].IsFirst, frame2.DataFrames[0].IsFirst);
            Assert.Equal(frame1.DataFrames[0].IsLast, frame2.DataFrames[0].IsLast);
            Assert.Equal(frame1.DataFrames[0].Header, frame2.DataFrames[0].Header);
            Assert.Equal(frame1.DataFrames[1].MessageNumber, frame2.DataFrames[1].MessageNumber);
            Assert.Equal(frame1.DataFrames[1].Offset, frame2.DataFrames[1].Offset);
            Assert.Equal(frame1.DataFrames[1].Length, frame2.DataFrames[1].Length);
            Assert.Equal(frame1.DataFrames[1].IsFirst, frame2.DataFrames[1].IsFirst);
            Assert.Equal(frame1.DataFrames[1].IsLast, frame2.DataFrames[1].IsLast);
            Assert.Equal(frame1.DataFrames[1].Header, frame2.DataFrames[1].Header);
            Assert.Equal(frame1.CancellationDetails, frame2.CancellationDetails);
        }

        /// <summary>
        /// Ensures a control frame respresenting a cancel (OpCode=0x12) can be serialized and deserialized
        /// </summary>
        [Fact]
        public void SerializeDeserializeCancel()
        {
            var frame1 = new ControlFrame()
            {
                OpCode = 0x12,
                RttEstimate = 42,
                ThroughputEstimate = 12345678,
                CancellationDetails = new MessageCancelControl()
                {
                    MessageNumbers = 42
                }
            };
            var bytes = frame1.Write();

            var frame2 = new ControlFrame(bytes);
            Assert.Equal(frame1.OpCode, frame2.OpCode);
            Assert.Equal(frame1.RttEstimate, frame2.RttEstimate);
            Assert.Equal(frame1.ThroughputEstimate, frame2.ThroughputEstimate);
            Assert.Equal(frame1.Capabilities, frame2.Capabilities);
            Assert.Equal(frame1.DataFrames, frame2.DataFrames);
            Assert.Equal(frame1.CancellationDetails.MessageNumbers, frame2.CancellationDetails.MessageNumbers);
        }
    }
}
